File: verysafe.py
"""This is not actually very safe.

This script is an attempt at generating as many true positives in Checkmarx
as possible. Where a specific defect is named, it does NOT guarantee that
Checkmarx will list it AS a defect. Named defects are best-effort attempts
at triggering a warning, nothing more.

The Checkmarx output should be compared against Snyk Code's output.
"""

import importlib
import logging
import os
import pickle
import random
import re
import socket
import subprocess
import sys

import psycopg2
import pymongo
import requests
from fastapi import FastAPI, Request

print("YOU SHOULD NOT RUN THIS COMMAND! YOU MIGHT BREAK SOMETHING!")
i = input("Enter I DO NOT CARE to ignore this warning:")
if i != "I DO NOT CARE":
  print("Good decision")
  sys.exit()

print("CODE/COMMAND INJECTION")
user_input = input("Something to execute:")

print("eval:", eval(user_input))
print("os.system:", os.system(user_input))
print("subprocess.Popen:", subprocess.Popen(user_input.split()))

if sys.argv:
  print("\nArgv:")
  args = " ".join(sys.argv)
  print("eval:", eval(args))
  print("os.system:", os.system(args))
  print("subprocess.Popen:", subprocess.Popen(sys.argv))

if os.path.exists(user_input):
  print("You gave me a file!")
  with open(user_input, "r") as f:
    contents = f.read()
    print("eval on file contents:", eval(contents))
    print("os.system on file contents:", os.system(contents))
    print("subprocess.Popen:", subprocess.Popen(contents.split()))

r = requests.get("https://dog.ceo/api/breeds/list/all")
r.raise_for_status()
dogs = r.json()

status = dogs["status"]
print("eval:", eval(status))
print("os.system:", os.system(status))
print("subprocess.Popen:", subprocess.Popen(status.split()))


def sanitize(stuff: str) -> str:
  return stuff


print("INJECTION WITH BAD SANITIZATION")
to_sanitize = input("Something to sanitize:")
sanitized = sanitize(to_sanitize)
print(to_sanitize)
print(sanitized)
print("eval:", eval(to_sanitize))
print("eval:", eval(sanitized))

sanitized = input("Enter something:")
print("eval:", eval(sanitized))


def desanitize(stuff):
  return stuff


desanitized = desanitize(to_sanitize)
print("eval:", eval(desanitized))

test = sanitize(3)  # Not sure if Checkmarx will or should complain; Snyk does not

print("REGEX-BASED SANITIZATION")
test = input("Give me some input:")
m = re.match(r"(.*)", test)  # Snyk accepts this line, which doesn't do anything, as sanitization
sanitized = m.group(1)
print("eval:", eval(test))
print("eval:", eval(sanitized))

print("HARDCODED ABSOLUTE PATH")
with open("/Users/c517600/Developer/verysafe/verysafe.py", "r") as f:
  print(f.read())

print("CONNECTION STRING INJECTION")
client = pymongo.MongoClient(user_input)

print("LOCAL FILE INCLUSION")
importlib.import_module(user_input)

print("OS ACCESS VIOLATION")
if os.path.exists(user_input):
  print("Deleting the user-input file")
  os.unlink(user_input)

print("SQL INJECTION")
db = client.test
db.collection.delete_one(eval(user_input))
db.collection.delete_one({"name": user_input})

# Alternate method
client["test"]["collection"].delete_one({"_id": user_input})

conn = psycopg2.connect(user_input)
cur = conn.cursor()
cur.execute(user_input)

print("UNSAFE DESERIALIZATION")
if os.path.exists(user_input):
  with open(user_input, "rb") as f:
    na = pickle.load(f)
    eval(na)

print("COMMUNICATION OVER HTTP and HARDCODED SECRET")
header = {"Authorization": "token AOriesnTIOEARNTIOENARIEST"}
r = requests.get("http://example.com", headers=header)

print("HARDCODED PASSWORD IN CONNECTION STRING")
r = requests.get("http://myfakeuser:AIOSENTAIREONestntrsenetrsnienrsNEINEIRST@example.com")

print("LOG SPOOFING")
logging.basicConfig(level=user_input)
logging.info(user_input)

print("INSECURE RANDOMNESS")
password = random.random() * 1000  # Should not be caught

print("RESOURCE INJECTION")
app = FastAPI()  # By not disabling the autodocs, this should probably also flag a warning


@app.api_route("/test", methods=["GET", "POST", "DELETE"])
async def do_something_unsafe(request: Request):
  try:
    port = int(request.args.get("PortNo"))
    server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    server_sock.bind(("", port))
    server_sock.listen(500)

    while True:
      (client_sock, address) = server_sock.accept()
      # Not sure if we need to actually do anything with this socket to trigger the defect

  except Exception:
    pass

File: verysafe.js
const express = require("express");
const { MongoClient } = require("mongodb");
const prompt = require("prompt-sync");
const readline = require("readline");
const fs = require("fs");
const { Client } = require("pg");
const { exec } = require("child_process");

const app = express();
const uri = prompt("Enter connection string:");
const client = new MongoClient(uri); // Connection string not sanitized
const db = client.testDb;
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

app.get("/search", (req, res) => {
  const results = db.coll.find(req.query.product);
  if (results.length === 0) {
    return res.send(`<p>No results found for "${req.query.product}"</p>`);
  }
  return res.send("<p>Something found</p>");
});

app.get("/delete", (req, res) => {
  db.coll.deleteOne({ _id: req.query.item });
  return res.send("Deleted!");
});

app.get("/eval-delete", (req, res) => {
  db.coll.deleteOne(eval(req.query.item));
  return res.send("Deleted!");
});

// Hardcoded secret
const secret = "ASIROTRSENTNnetnternisienrt78";

// Secret in connection string
const newClient = MongoClient("mongodb://myuser:AIESNTEIRSNenteternEE@example.com:27017");

// Unsafe eval() using external library
eval(prompt("Give me some JS to run! It's safe, I promise!"));

// Unsafe eval() using internal library
rl.question("What JS should I run?", input => {
  eval(input);
});

// Path traversal
rl.question("Give me a file. I'll see if it exists.", filename => {
  if (fs.existsSync(filename)) {
    console.log("It exists. Deleting");
    fs.unlinkSync(filename);
  } else {
    console.log("It does not exist");
  }
});

// SQL injection
pgClient = new Client();
pgClient.connect();

rl.question("What SQL should I execute?", sql => {
  pgClient.query(sql, null, (err, res) => {
    console.log(err ? err.stack : res.rows[0].message);
    pgClient.end();
  });
});

// Local file inclusion
rl.question("What should I include?", toInclude => {
  thing = require(toInclude);
});

// Code injection from command line arguments
const arg = process.argv.slice(2);
eval(arg[0]);

// Command injection from command line arguments
exec(arg[0], (error, stdout, stderr) => {
  console.log("This really should cause the tool to complain");
});

// Unsanitized connection string using CLI arguments
pgClient = new Client(arg[0]);